### ISSUE TITLE
What is the actual issue with python-pypdf?

### ISSUE THREAD

**Issue by alerque** [rrthomas/psutils#67]
I'm the Arch Linux packager for both `python-pypdf` and `psutils`, as well as a user of both for (non profit) book production work. Obviously some snafu has hit because `psbook` doesn't work at all:

Get ourselves a 4 page document to test with:

```console
$ sile -o four.pdf - <<< "<document><pagebreak /><pagebreak /><pagebreak /></document>"
SILE v0.15.4 (LuaJIT 2.1.1700206165) [Rust]
<STDIN> as xml
[1] [2] [3] [4]
````

Attempt to re-order pages for booklet layout:

```console
$ psbook four.pdf /dev/stdout
Traceback (most recent call last):
  File "/usr/bin/psbook", line 8, in <module>
    sys.exit(psbook())
             ^^^^^^^^
  File "/usr/lib/python3.12/site-packages/psutils/command/psbook.py", line 58, in psbook
    with file_transform(
  File "/usr/lib/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/psutils/transformers.py", line 528, in file_transform
    yield document_transform(
          ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/psutils/transformers.py", line 508, in document_transform
    return PdfTransform(indoc, outfile, size, in_size, specs, draw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/psutils/transformers.py", line 384, in __init__
    self.writer = PdfWriter(self.outfile)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pypdf/_writer.py", line 228, in __init__
    clone_from = PdfReader(clone_from)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pypdf/_reader.py", line 137, in __init__
    self.read(stream)
  File "/usr/lib/python3.12/site-packages/pypdf/_reader.py", line 571, in read
    self._basic_validation(stream)
  File "/usr/lib/python3.12/site-packages/pypdf/_reader.py", line 630, in _basic_validation
    stream.seek(0, os.SEEK_SET)
io.UnsupportedOperation: File or stream is not seekable.
```

I see in the release notes that pypdf was pinned to an older version because of failing tests, but I looked in the [pypdf issue tracker](https://github.com/py-pdf/pypdf/issues?q=) but don't see anything relevant raised upstream. In the mean time pinning to an older version is not generally a viable approach for distro packages because each package does not get a vendored version of all its dependencies and we can't hold packages back for everybody. Now that it has already gone out (before the psutils update even landed) it's not an easy proposition to downgrade it.

This means I kind of need to get to the bottom of what is actually wrong and fix it, even if that means backporting a patch to whatever is actually broken. Do you have any idea what the actual root problem is?

**Comment by alerque** [comment]
It looks like I missed [this issue on pypdf](https://github.com/py-pdf/pypdf/issues/2771) which is directly related to this.

**Comment by alerque** [comment]
I think there might be even more to the story. I've hacked together a system using archive packages such that I have pypdf 4.2.0 now, and the situation is definitely different. I had trouble piping `psbook <input> /dev/stdout | psnup /dev/stdin` with errors about an empty stream, so I decided to come back to this MWE but use a file instead:

```console
$ psbook four.pdf four-book.pdf
[4] [1] [2] [3]
Wrote 4 pages
```

That sure *sounds* successful, but is it? No the file is actually empty:

```console
$ stat four-book.pdf
stat four-book.pdf
  File: four-book.pdf
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 0,46    Inode: 208019409   Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   caleb)   Gid: ( 1000/   caleb)
Access: 2024-07-24 15:37:55.152208228 +0300
Modify: 2024-07-24 15:38:58.872427388 +0300
Change: 2024-07-24 15:38:58.872427388 +0300
 Birth: 2024-07-24 15:37:36.015476057 +0300
 ```

 This is with v3.3.2 of psutils.

**Comment by alerque** [comment]
Okay my issue with v3.3.2 above producing an empty PDF might be related to puremagic. Using v3.3.3 and paring it with hacked up parts from pypdf 4.2.0 does work. The puremagic update seems to have silently broken older versions of psutils. Silent failures are not a good thing in general but in this case dealing with distro packages its not that a big deal as long as current versions work — we just need to work out the compatibility with current pypdf versions.

**Comment by rrthomas** [comment]
I am a bit confused by this issue. There is no problem with pypdf 4.2.0 that I'm aware of. I linked to the [problem with 4.3.x](https://github.com/py-pdf/pypdf/issues/2771) in the commit where I pinned the version of pypdf: 801bb30.

**Comment by rrthomas** [comment]
I can't reproduce this problem with 3.3.3:

```
sile -o four.pdf - <<< "<document><pagebreak /><pagebreak /><pagebreak /></document>"
…
$ psbook four.pdf four-book.pdf
$ stat four-book.pdf
$ ls -s four-book.pdf 
12K four-book.pdf
```

This is with the latest puremagic. So, it sounds like some recent update in puremagic broke older versions, but luckily I put out a new release yesterday because I noticed that puremagic has just started shipping type definitions.

Hence, closing this issue, but of course feel free to open a new one if you find something wrong.

**Comment by rrthomas** [comment]
BTW, sile sounds interesting!

**Comment by alerque** [comment]
Begging your pardon, but I don't think this issue is resolved. The current version of psutils does not work when paired with the latest release of pypdf. I realize you have pinned and older version for the case where people use it with vendored dependencies, but given that is not always possible (e.g. the case of Linux distros that package this as a tool) I think this project is currently broken and there should be a tracking issue for it getting fixed. I realise there is a cross-linked issue for the upstream change that instigated this breakage, but reading that it sounds like it might be up to this project to adapt.

I got quite distracted with other comments here because I initially tried to work around this by downgrading psutils, then read abou the pypdf issue and tried to downgrade that. That's when it became apparent that puremagic is also involved in the mess, so that would have to be downgraded too. Using a current psutils and puremagic and downgrading pypdf does work. But that isn't a long term solution I can just move on from. The psutils package is currently not usable on Arch Linux unless. To use it people have to manually downgrade from an archived pypdf. Unless pypdf is going to revert their API change it seems like the fix needs to be on this side.

**Comment by rrthomas** [comment]
> Begging your pardon, but I don't think this issue is resolved. The current version of psutils does not work when paired with the latest release of pypdf. I realize you have pinned and older version for the case where people use it with vendored dependencies, but given that is not always possible (e.g. the case of Linux distros that package this as a tool) I think this project is currently broken and there should be a tracking issue for it getting fixed. I realise there is a cross-linked issue for the upstream change that instigated this breakage, but reading that it sounds like it might be up to this project to adapt.

I'm sorry, but fundamentally I am writing a Python package that is distributed via PyPI, and beyond that, packaging is Somebody Else's Problem (here, yours).

I am aware of the problem with PyPDF, and as I've said I have already been given a workaround, and if I don't get more information soon, I will make a release anyway.

Fundamentally there's not a lot I can do about compatibility with fast-moving dependencies. If PyPDF releases a new major version, then of course I don't expect to release a new version of PSPDFUtils immediately, though I certainly hope to before too long. I have had several experiences with PyPDF updates breaking the test suite of PSPDFUtils, which I put a lot of effort into resolving by accepting pixel-perfect results with a warning when they are not byte-identical. But fundamentally, I am shipping an application that uses a library, not a library that has to work with other libraries. If PyPDF breaks PSPDFUtils too often or evolves too fast for my release schedule, that's something that packagers of PSPDFUtils will have to deal with.

**Comment by rrthomas** [comment]
I've just released 3.3.4, which fixes compatibility with pyPDF 4.3.x.
